plugins/hls-overloaded-record-dot-plugin/src/Ide/Plugin/OverloadedRecordDot.hs

```diff
@@ -0,0 +1,240 @@
+
+instance NFData GhcExtension where
+  rnf x = x `seq` ()
+
+data RecordSelectors
+  = RecordSelectors RealSrcSpan (HsExpr (GhcPass 'Renamed))
+
+instance Pretty RecordSelectors where
+  pretty (RecordSelectors ss e) = pretty (printOutputable ss) <> ":" <+> pretty (printOutputable e)
+
+data ConvertedRecordSelector = ConvertedRecordSelector
+  { range              :: Range
+  , convertedDotRecord :: Text
+  }
+  deriving (Generic)
+
+instance Pretty ConvertedRecordSelector where
+  pretty (ConvertedRecordSelector r cdr) = pretty (show r) <> ":" <+> pretty cdr
+
+instance NFData ConvertedRecordSelector
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId)
+  { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionProvider
+  , pluginRules = collectConvRecSelsRule recorder
+  }
+
+codeActionProvider :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionProvider ideState pId (CodeActionParams _ _ caDocId caRange _) = pluginResponse $ do
+  nfp <- getNormalizedFilePath (caDocId ^. L.uri)
+  pragma <- getFirstPragma pId ideState nfp
+  CCRSR crsMap (map unExt -> exts) <- collectConvRecSels' ideState nfp
+  let actions = map (mkCodeAction nfp exts pragma) (RangeMap.filterByRange caRange crsMap)
+  pure $ List actions
+  where
+    mkCodeAction :: NormalizedFilePath -> [Extension] -> NextPragmaInfo -> ConvertedRecordSelector -> Command |? CodeAction
+    mkCodeAction nfp exts pragma crs = InR CodeAction
+      { _title = mkCodeActionTitle exts
+      , _kind = Just CodeActionRefactorRewrite
+      , _diagnostics = Nothing
+      , _isPreferred = Nothing
+      , _disabled = Nothing
+      , _edit = Just $ mkWorkspaceEdit nfp edits
+      , _command = Nothing
+      , _xdata = Nothing
+      }
+      where
+        edits = mkTextEdit crs : maybeToList pragmaEdit
+
+        mkTextEdit :: ConvertedRecordSelector -> TextEdit
+        mkTextEdit (ConvertedRecordSelector r cdr) = TextEdit r cdr
+
+        pragmaEdit :: Maybe TextEdit
+        pragmaEdit = if OverloadedRecordDot `elem` exts
+                       then Nothing
+                       else Just $ insertNewPragma pragma OverloadedRecordDot
+
+    mkWorkspaceEdit :: NormalizedFilePath -> [TextEdit] -> WorkspaceEdit
+    mkWorkspaceEdit nfp edits = WorkspaceEdit changes Nothing Nothing
+      where
+        changes = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List edits)
+
+    mkCodeActionTitle :: [Extension] -> Text
+    mkCodeActionTitle exts =
+      if OverloadedRecordDot `elem` exts
+        then title
+        else title <> " (needs extension: OverloadedRecordDot)"
+        where
+          title = "Convert to record dot syntax"
+
+collectConvRecSelsRule :: Recorder (WithPriority Log) -> Rules ()
+collectConvRecSelsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecordSelectors nfp ->
+  use TypeCheck nfp >>= \case
```

Comment:
I think if we use `useWithStale` with the `TypeCheck` rule, we shouldn't need to do so with `CollectRecordSelectors`, the reason being that `CollectRecordSelectors` should always be able to return a result, because it is no longer dependent on being able to get fresh `TypeCheck` data.

Reply:
Please consider not using `useWithStale` because this plugin is for code action, we don't have any necessary to keep stale results with the potential risk that the application may be messed up somehow.

Reply:
I don't think it's too bad, right? We have the position mapping so we can create our code actions and have them edit the right place. To make things go wrong I think you would need to:
1. Make the module not typecheck
2. Edit a field selector application to make it not a field selector application any more
3. Try and apply the rewriting code action

Even then, I think this might actually fail at the position mapping step. (And failing here just means "no code action is returned", which is okay-ish)

Reply:
FYI, rust-analyzer and typescript language server disabled most code actions if parse failed. *most* means I really see some code actions still work in rust-analyzer.